### PR TITLE
fix: add defensive nil checks during msg type casting

### LINKED ISSUE #1714: Defensive nil checks in the propagation reactor
we should add defensive nil checks to the propagation reactor,  and ensure that we trace each instance of hitting nil to ensure that they don't occur in the first place

### PR DESCRIPTION
**Pull request by ninabarbakadze**
## Description

Fixes #1714 

## Testing

- [x] Unit test for testing nil messages
- [x] Fuzz test for manipulating message fields with very high probability of nil values

Also went through propagation interface functions getting called in consensus a few times and tried to verify that no empty/nil structs were getting set.

### CODE REVIEW

File: consensus/propagation/types/types_fuzz_test.go [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+	"github.com/tendermint/tendermint/proto/tendermint/crypto"
+	protobits "github.com/tendermint/tendermint/proto/tendermint/libs/bits"
+	"github.com/tendermint/tendermint/proto/tendermint/propagation"
+	prototypes "github.com/tendermint/tendermint/proto/tendermint/types"
+	"github.com/tendermint/tendermint/types"
+)
+
+// TestFuzzMsgFromProto tests the MsgFromProto function with expected message types
+// and mutates them with high probability of nil values
+func TestFuzzMsgFromProto(t *testing.T) {
```
Comment (evan-forbes) [inline consensus/propagation/types/types_fuzz_test.go:18]: how long does this test take? do we need to run this every run? 

imo, if we know of an edge case and its easy enough to include in the correctness test, then we should add that

**Review by evan-forbes — APPROVED** [review]
I'm also fine with merging now and adding a flup since the tests are in the fuzzer, but will defer to the author

File: consensus/propagation/types/types.go [hunk 2]
```diff
@@ -450,6 +454,9 @@ func MsgFromProto(p *protoprop.Message) (Message, error) {
 			Prove:  msg.Prove,
 		}
 	case *protoprop.RecoveryPart:
+		if msg == nil {
+			return nil, errors.New("propagation: nil RecoveryPart")
+		}
```
[the remark this answers is not in the document]
Comment (ninabarbakadze) [reply consensus/propagation/types/types.go:459]: protobuf automatically converts nil byte slices to empty slices ([]byte{}) during unmarshaling.

File: consensus/propagation/types/types_test.go [hunk 3]
```diff
@@ -479,6 +479,46 @@ func TestHaveParts_GetIndex(t *testing.T) {
...
+			msg: &protoprop.Message{
+				Sum: &protoprop.Message_CompactBlock{},
+			},
+		},
+		{
+			name: "nil message: HaveParts",
+			msg: &protoprop.Message{
+				Sum: &protoprop.Message_HaveParts{},
+			},
+		},
```
[the remark this answers is not in the document]
Comment (ninabarbakadze) [reply consensus/propagation/types/types_test.go]: https://github.com/celestiaorg/celestia-core/pull/1772/commits/2c2f3f3cce240dae448302b23152275937d92931

File: consensus/propagation/types/types_test.go [hunk 4]
```diff
@@ -479,6 +479,46 @@ func TestHaveParts_GetIndex(t *testing.T) {
...
+	testCases := []struct {
+		name string
+		msg  *protoprop.Message
+	}{
+		{
+			name: "nil message: CompactBlock",
+			msg: &protoprop.Message{
+				Sum: &protoprop.Message_CompactBlock{},
+			},
+		},
```
[the remark this answers is not in the document]
Comment (ninabarbakadze) [reply consensus/propagation/types/types_test.go]: https://github.com/celestiaorg/celestia-core/pull/1772/commits/2c2f3f3cce240dae448302b23152275937d92931

File: consensus/propagation/types/types.go [hunk 5]
```diff
@@ -395,15 +395,15 @@ func MsgFromProto(p *protoprop.Message) (Message, error) {
...
-	case *protoprop.TxMetaData:
-		pb = &TxMetaData{
-			Hash:  msg.Hash,
-			Start: msg.Start,
-			End:   msg.End,
-		}
 	case *protoprop.CompactBlock:
+		if msg == nil {
+			return nil, errors.New("propagation: nil compact block")
+		}
```
[the remark this answers is not in the document]
Comment (ninabarbakadze) [reply consensus/propagation/types/types.go]: We had duplicate logic in `MsgFromProto` where we could've just used the methods that were already written for converting from proto to native go types like `CompactBlockFromProto`, etc.

**Review by evan-forbes — APPROVED** [review]
dope dope dope thanks
